apis/configwatch.go

```diff
@@ -181,3 +208,103 @@ func MonitorTLS(start time.Time, expiry time.Time, conf *config.Config) {
 		}
 	}
 }
+
+func (s *Server) getClientValidator(helloInfo *tls.ClientHelloInfo) func([][]byte, [][]*x509.Certificate) error {
+
+	log.Debug().Msgf("Inside get client validator - %v", helloInfo.Conn.RemoteAddr())
+	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		// Verifying client certs with root ca
+		opts := x509.VerifyOptions{
+			Roots:         s.CaCertPool,
+			CurrentTime:   time.Now(),
+			Intermediates: x509.NewCertPool(),
+			KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		}
+		_, err := verifiedChains[0][0].Verify(opts)
+		if err != nil {
+			log.Error().Err(err).Msgf("certs verification failed")
+			return err
+		}
+
+		log.Debug().Msgf("validating with SAN match rules - %s", s.SANMatchRules)
+		if len(s.SANMatchRules) == 0 {
+			log.Info().Msgf("mtls san match rules are undefined")
+			return nil
+		}
+		for _, dnsName := range verifiedChains[0][0].DNSNames {
+			dnsName := toLowerCaseASCII(dnsName)
```

Comment:
we need to not only lower case the dnsName from the SAN, but also validate it is valid according to the input.  Verify(opts) above won't do this since we are not relying on that for DNSName validation.  

Specifically we should do validation similar to this below and skip the DNS Name if it is not valid according to RFC.

https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/crypto/x509/verify.go;l=959;drc=d8220ac32fc73e69aa6d58890b6b5277d277825f